src/JuliusSweetland.OptiKey.Core/Services/KeyboardOutputService.cs

```diff
@@ -490,12 +491,66 @@ public void ProcessSingleKeyText(string capturedText)
         {
-            // TODO: This is a stub for future use
-            throw new NotImplementedException();            
+            Log.InfoFormat("Processing input value {0} with key press type {1}", inKey, type);
+
+            FunctionKeys funKey;
+            var virtualKeyCode = (Enum.TryParse(inKey, out funKey)) ? funKey.ToVirtualKeyCode() : null;
+
+            if (virtualKeyCode != null)
+            {
+                if (type == KeyPressKeyValue.KeyPressType.Press)
+                    publishService.KeyDown(virtualKeyCode.Value);
+                else if (type == KeyPressKeyValue.KeyPressType.Release)
+                    publishService.KeyUp(virtualKeyCode.Value);
+                else if (type == KeyPressKeyValue.KeyPressType.Toggle)
+                    publishService.KeyToggle(virtualKeyCode.Value);
+                else
+                    publishService.KeyDownUp(virtualKeyCode.Value);
+            }
+            else
+            {
+                //Get keyboard layout of currently focused window
+                IntPtr hWnd = PInvoke.GetForegroundWindow();
+                int lpdwProcessId;
+                int winThreadProcId = PInvoke.GetWindowThreadProcessId(hWnd, out lpdwProcessId);
+                IntPtr keyboardLayout = PInvoke.GetKeyboardLayout(winThreadProcId);
+
+                //Convert this into a culture string for logging
+                string keyboardCulture = "Unknown";
+                var installedInputLanguages = InputLanguage.InstalledInputLanguages;
+                for (int i = 0; i < installedInputLanguages.Count; i++)
+                {
+                    if (keyboardLayout == installedInputLanguages[i].Handle)
+                    {
+                        keyboardCulture = installedInputLanguages[i].Culture.DisplayName;
+                        break;
+                    }
+                }
+                foreach (var chaKey in inKey)
+                {
+                    var vkKeyScan = PInvoke.VkKeyScanEx(chaKey, keyboardLayout);
+                    var vkCode = vkKeyScan & 0xff;
+                    if (vkKeyScan != -1)
+                    {
+                        if (type == KeyPressKeyValue.KeyPressType.Press)
+                            publishService.KeyDown((VirtualKeyCode)vkCode);
+                        else if (type == KeyPressKeyValue.KeyPressType.Release)
+                            publishService.KeyUp((VirtualKeyCode)vkCode);
+                        else if (type == KeyPressKeyValue.KeyPressType.Toggle)
+                            publishService.KeyToggle((VirtualKeyCode)vkCode);
+                        else
+                            publishService.KeyDownUp((VirtualKeyCode)vkCode);
+                    }
+                    else
+                    {
+                        Log.InfoFormat("No virtual key code found for '{0}'", chaKey);
+                    }
+                }
+            }
```

Comment:
I was really unsatisfied with trying to force KeyDown/Up/Toggle logic into the existing methods, so I isolated them here. These unique actions have special needs and I'm betting some changes yet to be identified.